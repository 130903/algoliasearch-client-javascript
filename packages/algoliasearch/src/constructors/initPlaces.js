/* eslint import/namespace: [2, { allowComputed: true }] */
// @flow

import * as placesMethods from '../methods/places';
import { createRequester } from 'algoliasearch-requester';
import attachParameters from './attachParameters';
// todo: use a real requester
import { fakeRequester as httpRequester } from '../testUtils';

import type { ApiKey, AppId } from 'algoliasearch';
import type { RequestMethod, RequesterOptions } from 'algoliasearch-requester';

export default function initPlaces(
  {
    apiKey = '',
    appID = 'places',
    requester: extraRequester,
    options,
  }: {
    apiKey?: ApiKey,
    appID?: AppId,
    requester?: RequestMethod,
    options?: RequesterOptions,
  } = {}
) {
  // if there's an appId, there should also be an apiKey
  // all other cases are invalid
  if (
    ((appID === 'places' || appID === '') && apiKey !== '') ||
    (appID !== 'places' && appID !== '' && apiKey === '')
  ) {
    /* eslint-disable prefer-rest-params */
    throw new Error(
      `Credentials not valid: you gave an API key (${apiKey}), and appID (${
        appID
      })

initPlaces(${[...arguments].map(arg => JSON.stringify(arg)).join(',')})`
    );
    /* eslint-enable prefer-rest-params */
  }
  const requester = extraRequester
    ? extraRequester
    : createRequester({ appID, apiKey, options, httpRequester });

  return {
    ...attachParameters(placesMethods, { requester }),
    requester,
  };
}
