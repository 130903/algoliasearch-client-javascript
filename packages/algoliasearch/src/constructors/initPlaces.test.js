import initPlaces from './initPlaces';

it('initPlaces throws when it has an appId', () => {
  expect(() => initPlaces()).not.toThrow();
  expect(() => initPlaces({})).not.toThrow();
  expect(() => initPlaces({ appID: 'pl-cool', apiKey: 'key' })).not.toThrow();

  expect(() =>
    initPlaces({ apiKey: 'some_key' })
  ).toThrowErrorMatchingSnapshot();
  expect(() =>
    initPlaces({ appID: 'some_app' })
  ).toThrowErrorMatchingSnapshot();
});

it('initPlaces contains the correct methods', () => {
  const places = initPlaces();
  expect(Object.keys(places)).toMatchSnapshot();
});

it('places.search() works 🐣', async () => {
  const places = initPlaces();
  const result = await places.search({ query: 'hello world' });
  expect(result).toMatchSnapshot();
});
