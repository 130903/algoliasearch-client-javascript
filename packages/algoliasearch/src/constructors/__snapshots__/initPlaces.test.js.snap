// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initPlaces contains the correct methods 1`] = `
Array [
  "getPlace",
  "getPlaces",
  "search",
  "requester",
]
`;

exports[`initPlaces throws when it has an appId 1`] = `
"Credentials not valid: you gave an API key (some_key), and appID (places)

initPlaces({\\"apiKey\\":\\"some_key\\"})"
`;

exports[`initPlaces throws when it has an appId 2`] = `
"Credentials not valid: you gave an API key (), and appID (some_app)

initPlaces({\\"appID\\":\\"some_app\\"})"
`;

exports[`places.search() works üê£ 1`] = `
Object {
  "body": Object {
    "params": Object {
      "query": "hello world",
    },
  },
  "connectTimeout": 1000,
  "method": "POST",
  "requestOptions": undefined,
  "timeout": 2000,
  "url": Object {
    "hostname": "places-dsn.algolia.net",
    "pathname": "/1/indexes/places/queryundefined",
  },
}
`;
